src/lib/ranking.mli

```diff
@@ -44,6 +44,105 @@ module One : sig
 
 end
 
+module Matrix : sig
+
+  type ('acc, 'target) t = {
+    (* info about judges *)
+    head : bool;
+    judges : 'target array;
+    (* info about targets *)
+    targets : 'target array;
+    (* artefacts (and bonus) for each target/judge pair *)
+    mutable missing_artefacts : int;
+    artefacts : Artefact.t option array array;
+    bonus : Bonus.t array;
+    (* accumulator for ranking algorithms *)
+    ranking_acc : 'acc array;
+    (* resulting ranking *)
+    ranks : 'target One.t;
+  }
+
+  (* accessors *)
+
+  val ranks : ('acc, 'target) t -> 'target One.t
+
+  val length : ('acc, 'target) t -> int
+
+  val width : ('acc, 'target) t -> int
+
+  val target :('acc, 'target) t -> i:int -> 'target
+
+  val judge : ('acc, 'target) t -> j:int -> 'target
+
+  val bonus : ('acc, 'target) t -> i:int -> Bonus.t
+
+  val head : ('acc, 'target) t -> int option
+
+  val is_head : ('acc, 'target) t -> j:int -> bool
+
+  val missing_artefacts : ('acc, 'target) t -> int
+
+  val artefact : ('acc, 'target) t -> i:int -> j:int -> Artefact.t option
+
+  (* debug printing *)
+
+  val printbox_matrix : acc_line:('acc -> PrintBox.t array) ->
+    acc_side:[< `Left | `Right ] ->
+    pp:(Format.formatter -> 'target -> unit) ->
+    ('acc, 'target) t ->
+    PrintBox.t array array
+
+  val printbox: acc_line:('acc -> PrintBox.t array) ->
+    acc_side:[< `Left | `Right ] ->
+    pp:(Format.formatter -> 'target -> unit) ->
+    ('acc, 'target) t ->
+    PrintBox.t
+
+  (* initialization *)
+
+  val init : head:'target option -> judges:'target list ->
+    targets: 'target list ->
+    acc:(int -> 'acc) -> ('acc, 'target) t
+
+  val map: targets:('a -> 'b) ->
+    judges:('a -> 'b) ->
+    ('c, 'a) t ->
+    ('c, 'b) t
+
+  val iteri: targets:(int -> 'a -> unit) ->
+    judges:(int -> 'a -> unit) ->
+    ('b, 'a) t ->
+    unit
+
+  (* filing up the matrix with artefacts *)
+
+  val acc_bonus : i:int -> bonus:int -> ('a, 'b) t -> unit
+
+  val acc_artefact : i:int -> j:int -> artefact:Artefact.t option -> ('a, 'b) t -> unit
+
+  (* Ranking helpers *)
+
+  val get : i:int -> ('a, 'b) t -> 'a
+
+  val set : i:int -> ('a, 'b) t -> 'a -> unit
+
+  val swap : ('a, 'b) t -> int -> int -> unit
+
+  val sort : cmp:('a -> 'a -> int) ->
+    start:int ->
+    stop:int ->
+    ('a, 'b) t ->
+    unit
+
+  val segments : cmp:('a -> 'a -> int) ->
+    start:int ->
+    stop:int ->
+    f:(start:int -> stop:int -> unit) ->
+    ('a, 'b) t ->
+    unit
+
+end
```

Comment:
The `Matrix` module was volontarily not exposed, why do you expose it ? If it is to access the ranking accumulators, then it would be better to expose dedicated types for the "explanation" of each ranking algorithm, and generate these explanations from the matrix internally.